pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,261 @@
+	}
+	return nil
+}
+
+//prepareLocalProfileServerMap process configuration of network instances
+//to find match with defined localServerURL
+func prepareLocalProfileServerMap(localServerURL string, getconfigCtx *getconfigContext) (map[string]net.IP, error) {
+	res := make(map[string]net.IP)
+
+	netInstanses := getconfigCtx.pubNetworkInstanceConfig.GetAll()
+	u, err := url.Parse(localServerURL)
+	if err != nil {
+		return nil, fmt.Errorf("checkAndPrepareLocalIP: url.Parse: %s", err)
+	}
+	localProfileServerHostname := u.Hostname()
+	localProfileServerIP := net.ParseIP(localProfileServerHostname)
+	for _, entry := range netInstanses {
+		config := entry.(types.NetworkInstanceConfig)
+		if localProfileServerIP != nil {
+			//defined ip
+			if config.Subnet.Contains(localProfileServerIP) {
+				res[localServerURL] = config.Gateway
+			}
+		} else {
+			//defined host, search in DNS records
+			for _, el := range config.DnsNameToIPList {
+				if el.HostName == localProfileServerHostname {
+					for _, ip := range el.IPs {
+						localServerURLReplaced := strings.Replace(localServerURL, localProfileServerHostname,
+							ip.String(), 1)
+						res[localServerURLReplaced] = config.Gateway
+						log.Functionf(
+							"prepareLocalProfileServerMap: will use url with replaced hostname: %s",
+							localServerURLReplaced)
+					}
+				}
+			}
+		}
+	}
+	return res, nil
+}
+
+// getLocalProfileConfig connects to local profile server to fetch current profile
+func getLocalProfileConfig(localServerURL string, iteration int, getconfigCtx *getconfigContext) error {
+
+	log.Tracef("getLocalProfileConfig(%s, %d)", localServerURL, iteration)
+
+	localServerMap, err := prepareLocalProfileServerMap(localServerURL, getconfigCtx)
+	if err != nil {
+		return fmt.Errorf("getLocalProfileConfig: prepareLocalProfileServerMap: %s", err)
+	}
+
+	if len(localServerMap) == 0 {
+		return fmt.Errorf(
+			"getLocalProfileConfig: cannot find any configured local subnets for localServerURL: %s",
+			localServerURL)
+	}
+
+	var errList []string
+	for link, srcIP := range localServerMap {
+		resp, contents, err := zedcloud.SendLocal(zedcloudCtx, link, srcIP, 0, nil)
+		if err != nil {
+			errList = append(errList, fmt.Sprintf("SendLocal: %s", err))
+			continue
+		}
+		if resp.StatusCode == http.StatusOK {
+			if err := validateProtoMessage(link, resp); err != nil {
+				// send something to ledmanager ???
+				errList = append(errList, fmt.Sprintf("validateProtoMessage: resp header error: %s", err))
+				continue
+			}
+			getconfigCtx.localProfileReceived = true
+			err := validateAndSetLocalProfile(contents, getconfigCtx)
+			if err != nil {
+				errList = append(errList, fmt.Sprintf("validateAndSetLocalProfile: %s", err))
+				continue
+			} else {
+				return nil
+			}
+		}
+		errList = append(errList, fmt.Sprintf("getLocalProfileConfig: wrong response status code: %d",
+			resp.StatusCode))
+	}
+	if !getconfigCtx.readSavedLocalProfile {
```

Comment:
Why do you need to re-read the saved proto file here? It should be sufficient to do that at startup.

Reply:
Good point, getting it from parseProfile first time we have config seems enough.

Reply:
Not from parseprofile but from the top of the task code.